airbyte-webapp-e2e-tests/cypress/integration/connection/syncModes.spec.ts

```diff
@@ -0,0 +1,379 @@
+      });
+
+      streamsTable.checkNoSourceDefinedCursor("public", "users");
+      streamsTable.checkNoSourceDefinedPrimaryKeys("public", "users");
+    });
+  });
+
+  describe("Incremental | Deduped + history", () => {
+    describe("with source-defined primary keys", () => {
+      const cursor = "updated_at";
+      const primaryKey = "id";
+
+      before(() => {
+        streamsTable.searchStream("users2");
+        streamsTable.selectSyncMode("Incremental", "Deduped + history");
+      });
+
+      it("should be able to select cursor", () => {
+        streamsTable.hasEmptyCursorSelect("public", "users2");
+        streamsTable.selectCursor("users2", cursor);
+        streamsTable.checkSelectedCursorField("users2", cursor);
+      });
+
+      it("has source-defined primary key", () => {
+        streamsTable.checkSourceDefinedPrimaryKeys("users2", primaryKey);
+      });
+
+      it("saves", () => {
+        saveConnectionAndAssertStreams({
+          namespace: "public",
+          name: "users2",
+          config: {
+            syncMode: "incremental",
+            destinationSyncMode: "append_dedup",
+            cursorField: [cursor],
+            primaryKey: [[primaryKey]],
+          },
+        });
+
+        streamsTable.checkSelectedCursorField("users2", cursor);
+        streamsTable.checkSourceDefinedPrimaryKeys("users2", primaryKey);
+      });
+    });
+
+    describe("with source-defined cursor and primary keys", () => {
+      const cursor = "updated_at";
+      const primaryKey = "id";
+
+      before(() => {
+        streamsTable.searchStream("accounts");
+        streamsTable.selectSyncMode("Incremental", "Deduped + history");
+      });
+
+      it("has source-defined cursor", () => {
+        streamsTable.checkSourceDefinedCursor("accounts", cursor);
+      });
+
+      it("has source-defined primary key", () => {
+        streamsTable.checkSourceDefinedPrimaryKeys("accounts", primaryKey);
+      });
+
+      it("saves", () => {
+        saveConnectionAndAssertStreams({
+          namespace: "public",
+          name: "accounts",
+          config: {
+            syncMode: "incremental",
+            destinationSyncMode: "append_dedup",
+            cursorField: ["updated_at"],
+            primaryKey: [["id"]],
+          },
+        });
+
+        streamsTable.checkSourceDefinedCursor("accounts", cursor);
+        streamsTable.checkSourceDefinedPrimaryKeys("accounts", primaryKey);
+      });
+    });
+
+    describe("with selectable primary keys", () => {
+      const cursorValue = "created_at";
+      const primaryKeyValue = ["car_id", "user_id"];
+
+      before(() => {
+        streamsTable.searchStream("user_cars");
+        streamsTable.selectSyncMode("Incremental", "Deduped + history");
+      });
+
+      it("has empty cursor and primary key selects", () => {
+        streamsTable.hasEmptyCursorSelect("public", "user_cars");
+        streamsTable.hasEmptyPrimaryKeySelect("public", "user_cars");
```

Comment:
💬 Discussion question:
Also faced the same flow - we always need to provide a namespace and stream name to have the stream we are looking for. And since in most of the cases we are working just with one stream, I decided to try to solve it via closure, for example:
```
const {toggleStreamSync, isStreamSyncEnabled, isStreamRowViisible} = getStreamUtitilyFunctions("namespace", "streamname")
```

Raw example - then, in tests we just need:

```
it("...",() =>{
    toggleStreamSync()
    isStreamSyncEnabled(true)
} )

it("...",() =>{
    toggleStreamSync()
    isStreamSyncEnabled(false)
} )
```

WDYT?

Reply:
I had some ideas on how to improve this but decided to stick with what we have available because the changes would also impact `connection.cy.ts`. I agree that it's better to just work within the context of the stream instead of calling the same thing repeatedly but that's something that I'd like to refactor later.

Reply:
Oh, okay, gotcha 👌